kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -117,307 +134,120 @@ public suspend inline fun <T> Mutex.withLock(owner: Any? = null, action: () -> T
-            if (!take()) return false
-            return cont.tryResume(Unit, idempotent = null) {
-                // if this continuation gets cancelled during dispatch to the caller, then release the lock
-                unlock(owner)
-            } != null
+    private inner class CancellableContinuationWithOwner(
+        val cont: CancellableContinuation<Unit>,
+        val owner: Any?
+    ) : CancellableContinuation<Unit> by cont {
+        override fun tryResume(value: Unit, idempotent: Any?, onCancellation: ((cause: Throwable) -> Unit)?): Any? {
+            val token = cont.tryResume(value, idempotent, onCancellation)
+            if (token !== null) this@MutexImpl.owner.value = owner
+            return token
         }
 
-        override fun completeResumeLockWaiter() = cont.completeResume(RESUME_TOKEN)
-        override fun toString(): String = "LockCont[$owner, ${cont}] for ${this@MutexImpl}"
+        override fun resume(value: Unit, onCancellation: ((cause: Throwable) -> Unit)?) {
+            this@MutexImpl.owner.value = owner
+            cont.resume(value, onCancellation)
+        }
     }
 
-    private inner class LockSelect<R>(
-        owner: Any?,
-        @JvmField val select: SelectInstance<R>,
-        @JvmField val block: suspend (Mutex) -> R
-    ) : LockWaiter(owner) {
-        override fun tryResumeLockWaiter(): Boolean = take() && select.trySelect()
-        override fun completeResumeLockWaiter() {
-            block.startCoroutineCancellable(receiver = this@MutexImpl, completion = select.completion) {
-                // if this continuation gets cancelled during dispatch to the caller, then release the lock
-                unlock(owner)
+    private inner class SelectInstanceWithOwner<Q>(
+        val select: SelectInstance<Q>,
+        val owner: Any?
+    ) : SelectInstance<Q> by select {
+        override fun trySelect(clauseObject: Any, result: Any?): Boolean {
+            return select.trySelect(clauseObject, result).also { success ->
+                if (success) this@MutexImpl.owner.value = owner
             }
         }
-        override fun toString(): String = "LockSelect[$owner, $select] for ${this@MutexImpl}"
-    }
-
-    // atomic unlock operation that checks that waiters queue is empty
-    private class UnlockOp(
-        @JvmField val queue: LockedQueue
-    ) : AtomicOp<MutexImpl>() {
-        override fun prepare(affected: MutexImpl): Any? =
-            if (queue.isEmpty) null else UNLOCK_FAIL
 
-        override fun complete(affected: MutexImpl, failure: Any?) {
-            val update: Any = if (failure == null) EMPTY_UNLOCKED else queue
-            affected._state.compareAndSet(this, update)
+        override fun selectInRegistrationPhase(internalResult: Any?) {
+            select.selectInRegistrationPhase(internalResult)
+            this@MutexImpl.owner.value = owner
         }
     }
+
+    override fun toString() = "Mutex@${hexAddress}"
```

Comment:
It would be nice to have meaningful `toString` -- e.g. adding whether it's locked and the owner to the string